extractor/image/image.go

```diff
@@ -0,0 +1,240 @@
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	layers := img.src.LayerInfos()
+
+	// ignore the cache error
+	_ = img.storeLayerInfosInCache(img.name, img.isFile, layers)
+
+	return layers, nil
+}
+
+func (img Image) getLayerInfosInCache(imageName string, isFile bool) ([]imageTypes.BlobInfo, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("layerinfos::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	var layerInfos []imageTypes.BlobInfo
+	if err := json.NewDecoder(rc).Decode(&layerInfos); err != nil {
+		return nil, xerrors.Errorf("invalid cache: %w", err)
+	}
+	return layerInfos, nil
+}
+
+func (img Image) storeLayerInfosInCache(imageName string, isFile bool, layers []imageTypes.BlobInfo) error {
+	// it doesn't store the information of a tar file
+	if isFile {
+		return nil
+	}
+
+	b, err := json.Marshal(layers)
+	if err != nil {
+		return xerrors.Errorf("invalid json: %w", err)
+	}
+	return img.cache.SetBytes("layerinfos::"+imageName, b)
+}
+
+func (img *Image) ConfigBlob(ctx context.Context) ([]byte, error) {
+	b, _ := img.getConfigBlobInCache(img.name, img.isFile)
+	if b != nil {
+		return b, nil
+	}
+
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	blob, err := img.src.ConfigBlob(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get a config: %w", err)
+	}
+
+	if !img.isFile {
+		_ = img.cache.SetBytes("configblob::"+img.name, blob)
+	}
+
+	return blob, nil
+}
+
+func (img Image) getConfigBlobInCache(imageName string, isFile bool) ([]byte, error) {
+	// tar file doesn't have info in cache
+	if isFile {
+		return nil, nil
+	}
+	rc := img.cache.Get("configblob::" + imageName)
+	if rc == nil {
+		return nil, nil
+	}
+	b, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to read a cache file: %w", err)
+	}
+	return b, nil
+}
+
+func (img *Image) GetBlob(ctx context.Context, dig digest.Digest) (io.ReadCloser, error) {
+	rc := img.cache.Get(dig.String())
+	if rc != nil {
+		return rc, nil
+	}
+
+	if err := img.populateSource(); err != nil {
+		return nil, xerrors.Errorf("failed population: %w", err)
+	}
+
+	rc, _, err := img.rawSource.GetBlob(ctx, imageTypes.BlobInfo{Digest: dig, Size: -1}, img.blobInfoCache)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
+	}
+
+	stream, _, err := compression.AutoDecompress(rc)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
+	}
+
+	r, _ := img.cache.Set(dig.String(), stream)
+
+	return ioutil.NopCloser(r), nil
+}
+
+func (img Image) RecordDigestUncompressedPair(dig digest.Digest, uncompressed digest.Digest) {
+	img.blobInfoCache.RecordDigestUncompressedPair(dig, uncompressed)
+}
```

Comment:
This doesn't seem to be used anywhere. Remove?

Reply:
OK. I will add it in the next PR.